test/functional/p2p_block_sync.py

```diff
@@ -0,0 +1,33 @@
+#!/usr/bin/env python3
+# Copyright (c) 2022 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test block download
+
+Ensure that even in IBD, we'll eventually sync chain from inbound peers
+"""
+
+from test_framework.test_framework import BitcoinTestFramework
+
+class BlockSyncTest(BitcoinTestFramework):
+
+    def set_test_params(self):
+        self.setup_clean_chain = True
+        self.num_nodes = 3
+
+    def setup_network(self):
+        self.setup_nodes()
+        self.connect_nodes(0, 1)
+        self.connect_nodes(1, 2)
+
+    def run_test(self):
+        self.log.info("Setup network: node0->node1->node2")
+        self.log.info("Mining one block on node0")
+        self.generate(self.nodes[0], 1)
+        self.log.info("Check that all nodes synced")
+        self.sync_all()
```

Comment:
nit: I think this can be removed, since `self.generate` already calls this internally (and fails if the sync fails).

It might also be good to explain why three nodes are needed and why node0->node1 isn't sufficient to test this.

My understanding is that this is setting up `node1` as the subject under test. Having an outbound and an inbound connection, of which the outbound is selected for block download, but fails to provide any data. The test would pass with just `node0->node1` setup which means that we already sync from inbound peers if there are no outbounds at all?

Reply:
I checked for the suggestions of @MarcoFalke, and made sure they were correct. In the definition of the `generate` function, `sync_all` is being called by default.
https://github.com/bitcoin/bitcoin/blob/master/test/functional/test_framework/test_framework.py#L633-L636
Also, I checked that the test works perfectly with just two nodes, 0 and 1.
I made the following changes to the test to incorporate the suggestions.

**Diff:**

```diff
     def set_test_params(self):
         self.setup_clean_chain = True
-        self.num_nodes = 3
+        self.num_nodes = 2
 
     def setup_network(self):
         self.setup_nodes()
         self.connect_nodes(0, 1)
-        self.connect_nodes(1, 2)
 
     def run_test(self):
-        self.log.info("Setup network: node0->node1->node2")
-        self.log.info("Mining one block on node0")
+        self.log.info("Setup network: node0->node1")
+        self.log.info("Mining one block on node0 and check that all nodes synced")
         self.generate(self.nodes[0], 1)
-        self.log.info("Check that all nodes synced")
-        self.sync_all()
         self.log.info("Success!")


```

Reply:
>  Also, I checked that the test works perfectly with just two nodes, 0 and 1.

Jup, the test with two nodes is passing on master and this pull. I think the goal of the test was to have a test that *fails* on master, but passes with this pull. For this you need three nodes.

Reply:
I just checked the test on the Master branch, and it seems like the test passes on the Master branch with two nodes. Since the original goal of this test was that it should fail on the Master as explained in PR's description. I think the number of nodes should be kept at 3.

We can still incorporate the following change to the test to remove redundancy from it:

Diff: 
```diff
def run_test(self):
         self.log.info("Setup network: node0->node1->node2")
-        self.log.info("Mining one block on node0")
+        self.log.info("Mining one block on node0 and check that all nodes synced")
         self.generate(self.nodes[0], 1)
-        self.log.info("Check that all nodes synced")
-        self.sync_all()
         self.log.info("Success!")

```

Reply:
> My understanding is that this is setting up node1 as the subject under test. Having an outbound and an inbound connection, of which the outbound is selected for block download, but fails to provide any data. The test would pass with just node0->node1 setup which means that we already sync from inbound peers if there are no outbounds at all?

@MarcoFalke Yes that is correct. We already had logic in place to download blocks from inbound peers in the case that we had no outbound peers, but prior to this patch we have had no way to download blocks that are only known by inbound peers if we're in IBD.

Reply:
I've updated the test comments and removed the `self.sync_all()` call.